### PR TITLE
Add pressure metric for bpf maps

### PR DESCRIPTION
**Pull request by jcaamano**
Adds a metric to tell the fill percentage from 0 to 1 of selected maps
filled from userspace. Selected maps are those non LRU filled from
userspace.

This metric is a Prometheus const label gauge that is registered and
unregistered dynamically for each map when over/under a threshold
respectively. By default the threshold is 0.0 for all tracked maps.

As other metrics, this one can be enabled or disabled globally via
command line and is disabled by default.

This metric takes advantage of the existing cache mechanism for maps
to count the current number of entries on each map. For those maps that
already have this cache enabled, there is no change in this regard. For
those that don't, this caching mechanism is partially enabled to track a
key-only cache, which is the case for ` cilium_lb4_reverse_sk`,
`cilium_lb6_reverse_sk`, `cilium_lb4_maglev` and `cilium_lb6_maglev`
maps.

The policy map is an exceptional case. For this map threshold is set to
0.1 to avoid flooding Prometheus with reports. The metric is tracked
from the endpoint instead of the map in this case as the endpoint is
already caching this data while the map does not.

**Release note**
```release-note
Add new `cilium_bpf_map_pressure` metric measuring the fill-up ratio of selected BPF maps.
```

### CODE REVIEW

File: pkg/bpf/map_linux.go [hunk 1]
```diff
@@ -214,6 +225,10 @@ func (m *Map) commonName() string {
 	return m.cachedCommonName
 }
 
+func (m *Map) nonPrefixedName() string {
+	return m.name[7:]
```
Comment (aanm) [inline pkg/bpf/map_linux.go]: and why `7`?
```
if len(m.name) >= 7 {
return m.name[7:]
}
return m.name
```

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/bpf/map_linux.go (hunk 1 above)
Comment (jcaamano) [reply pkg/bpf/map_linux.go]: Removing `cilium_` prefix but unlike it is done with the common name we want to keep things like the endpoint id in the name.

File: pkg/bpf/map_linux.go [hunk 2]
```diff
@@ -117,7 +117,10 @@ type Map struct {
 	// DumpParser is a function for parsing keys and values from BPF maps
 	DumpParser DumpParser
 
-	cache map[string]*cacheEntry
```
Comment (vadorovsky) [inline pkg/bpf/map_linux.go:120]: I don't think you have to change the type of `cache`. For empty entries we can just use `nil` (with an explicit assignment, like you already do, so the key would exists and we could look up for it). And we could avoid type conversions.

Comment (jcaamano) [reply pkg/bpf/map_linux.go:120]: Right, I was using a different empty struct type before but forgot to change this back.

File: pkg/bpf/map_linux.go (hunk 1 above)
Comment (joestringer) [reply pkg/bpf/map_linux.go]: I would suggest using something like `strings.TrimPrefix(...)` to prevent unintended behaviour here.

File: pkg/metrics/metrics.go [hunk 3]
```diff
@@ -553,6 +555,7 @@ func DefaultMetrics() map[string]struct{} {
 		Namespace + "_" + SubsystemKVStore + "_quorum_errors_total":                  {},
 		Namespace + "_fqdn_gc_deletions_total":                                       {},
 		Namespace + "_" + SubsystemBPF + "_map_ops_total":                            {},
+		Namespace + "_" + SubsystemBPF + "_map_pressure":                             {},
```
Comment (vadorovsky) [inline pkg/metrics/metrics.go]: To make it optional, maybe let's remove it from here and apply the following Helm change I made locally:

https://github.com/mrostecki/cilium/commit/bcd661b90683ace4f637e4bf2b51b962f7d2a895

Feel free to just pull/am/cherry-pick from my branch, if my changes look good to you (or modify them if they don't), there is also a Grafana change:

https://github.com/mrostecki/cilium/commits/map-pressure-metric

**Comment by borkmann** [comment]
@jcaamano @mrostecki Btw, something that might be of interest to you wrt mem usage: https://lore.kernel.org/bpf/20201125030119.2864302-1-guro@fb.com/ . See my comment here: https://lore.kernel.org/bpf/ef140167-8d80-c581-318c-36c0430e4cfa@iogearbox.net/ If we could use memcg to account the used memory, that would solve your issue as well (assuming it's always accounted to the memcg where the map was created in).

File: pkg/metrics/metrics.go (hunk 3 above)
Comment (jcaamano) [reply pkg/metrics/metrics.go]: I removed the metric from the defaults and added your commits to this PR. The helm one might have merit for it's own PR though.

**Comment by jcaamano** [comment]
> @jcaamano @mrostecki Btw, something that might be of interest to you wrt mem usage: https://lore.kernel.org/bpf/20201125030119.2864302-1-guro@fb.com/ . See my comment here: https://lore.kernel.org/bpf/ef140167-8d80-c581-318c-36c0430e4cfa@iogearbox.net/ If we could use memcg to account the used memory, that would solve your issue as well (assuming it's always accounted to the memcg where the map was created in, hence my discussion on the matter).

I see how this can help but might not be able to leverage it any time soon, right?. I guess that if the bpf maps can be sized independently this metric could still make sense as represented even if in the future we change the implementation of how we are measuring it.

**Comment by mrostecki** [comment]
> @jcaamano @mrostecki Btw, something that might be of interest to you wrt mem usage: https://lore.kernel.org/bpf/20201125030119.2864302-1-guro@fb.com/ . See my comment here: https://lore.kernel.org/bpf/ef140167-8d80-c581-318c-36c0430e4cfa@iogearbox.net/ If we could use memcg to account the used memory, that would solve your issue as well (assuming it's always accounted to the memcg where the map was created in, hence my discussion on the matter).

That looks good and indeed helpful for our case, I will take a deeper look. I think the question is - should we proceed with this PR as a temporary solution before BPF adopts memcg (and Cilium adopts that change by managing cgroups), or should we rather just wait for those memcg patches to get in and eventually backport them (f.e. to SUSE kernel)?

File: install/kubernetes/cilium/values.yaml [hunk 4]
```diff
@@ -863,6 +863,10 @@ prometheus:
     # -- Specify the Kubernetes namespace where Prometheus expects to find
     # service monitors configured.
     # namespace: ""
+  # Metrics that should be enabled or disabled from the default metric
```
Comment (joestringer) [inline install/kubernetes/cilium/values.yaml]: Couple of things here:

nit, formatting like this (like the other options above) allows the automated tooling we use on these values to pick up the descriptions and properly format them:
```suggestion
  # -- Metrics that should be enabled or disabled from the default metric
```

Additionally, would it be reasonable to also document what the valid options are here? Or is that better documented somewhere else?

File: pkg/metrics/metrics_test.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
+// Copyright 2019 Authors of Cilium
```
Comment (qmonnet) [inline pkg/metrics/metrics_test.go]: Nit: Copyright year should be 2020?

File: pkg/metrics/metrics_test.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	c.Assert(err, IsNil)
+	c.Assert(metrics, HasLen, initMetricLen+1)
+	c.Assert(GetGaugeValue(gauge.gauge), Equals, overThreshold)
+
+	gauge.Set(underThreshold)
+	metrics, err = registry.Gather()
+	c.Assert(err, IsNil)
+	c.Assert(metrics, HasLen, initMetricLen)
+	c.Assert(GetGaugeValue(gauge.gauge), Equals, underThreshold)
+
```
Comment (qmonnet) [inline pkg/metrics/metrics_test.go]: Nit: unnecessary blank line.

**Review by qmonnet — CHANGES_REQUESTED** [review]
Changes to the documentation look good.

I only had a quick glance at the changes in the agent - Please find some minor nits below.

**Comment by mrostecki** [comment]
test-me-please

**Review by qmonnet — APPROVED** [review]
Thanks a lot!

**Review by kaworu — APPROVED** [review]
Helm related change LGTM

File: pkg/bpf/map_linux.go [hunk 7]
```diff
@@ -245,11 +257,51 @@ func (m *Map) scheduleErrorResolver() {
...
 }
 
+// WithPressureMetricThreshold enables the tracking of a metric that measures
+// the pressure of this map. This metric is only reported if over the
+// threshold.
+func (m *Map) WithPressureMetricThreshold(threshold float64) *Map {
+	// When pressure metric is enabled, we keep track of map keys in cache
+	if m.cache == nil {
+		m.cache = map[string]*cacheEntry{}
+	}
```
Comment (aanm) [inline pkg/bpf/map_linux.go:277]: Maybe would be better to call `m.WithCache()` here? otherwise we will have `m.cache` with data but `m.withCache` will be false.

File: pkg/endpoint/bpf.go [hunk 8]
```diff
@@ -1071,6 +1073,19 @@ func (e *Endpoint) SkipStateClean() {
...
 }
 
+func (e *Endpoint) updatePolicyMapPressureMetric() {
+	if !option.Config.MetricsConfig.BPFMapPressure {
+		return
+	}
+
+	if e.policyMapPressureGauge == nil {
+		e.policyMapPressureGauge = metrics.NewBPFMapPressureGauge(e.policyMap.NonPrefixedName(), policymap.PressureMetricThreshold)
+	}
```
Comment (aanm) [inline pkg/endpoint/bpf.go]: This is something that can potentially be in the endpoint constructor.

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/bpf/map_linux.go (hunk 7 above)
Comment (jcaamano) [reply pkg/bpf/map_linux.go:277]: `m.cache` and `m.WithCache()`/`m.withCache` are different things. `m.cache` is used also for the pressure metric key-only cache unless `m.withCache` flag is set in which case is a key-value cache.  Perhaps renaming `m.withCache` to `m.withValueCache` is a bit more clear.

File: pkg/endpoint/bpf.go (hunk 8 above)
Comment (jcaamano) [reply pkg/endpoint/bpf.go]: Yes. I wasn't sure if there was a single endpoint constructor or many. Let me try to track them down, please check after.

Comment (jcaamano) [reply pkg/endpoint/bpf.go]: I moved this to be constructed with the policyMap which is deferred from the endpoint constructor as well but at least we are not checking for it on each metric value set.

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
retest-net-next

**Comment by mrostecki** [comment]
retest-4.9

File: pkg/bpf/map_linux.go [hunk 9]
```diff
@@ -880,6 +938,8 @@ func (m *Map) deleteCacheEntry(key MapKey, err error) {
 	k := key.String()
 	if err == nil {
 		delete(m.cache, k)
+	} else if !m.withValueCache {
+		return
```
Comment (joestringer) [inline pkg/bpf/map_linux.go:946]: Why is this necessary? Line 934 already checks `if m.cache == nil { return }`.

File: pkg/bpf/map_linux.go [hunk 10]
```diff
@@ -117,6 +118,9 @@ type Map struct {
 	// DumpParser is a function for parsing keys and values from BPF maps
 	DumpParser DumpParser
 
+	// withValueCache is true when map cache has been enabled
+	withValueCache bool
```
Comment (joestringer) [inline pkg/bpf/map_linux.go:122]: Is this different than `m.cache != nil`? Tracking an extra variable seems like provides an opportunity for the bool vs. the pointer to get out of sync.

If there is an actual difference, that would be useful context to put in the comment above this field.

File: pkg/bpf/map_linux.go [hunk 11]
```diff
@@ -933,7 +994,7 @@ func (m *Map) DeleteAll() error {
 
 	nextKey := make([]byte, m.KeySize)
 
-	if m.cache != nil {
+	if m.withValueCache {
```
Comment (joestringer) [inline pkg/bpf/map_linux.go:1006]: I couldn't figure out, does `DeleteAll()` update the metric?

File: pkg/bpf/map_linux.go [hunk 12]
```diff
@@ -836,22 +888,28 @@ func (m *Map) Update(key MapKey, value MapValue) error {
 	m.lock.Lock()
 	defer m.lock.Unlock()
 
+	defer m.updatePressureMetric()
```
Comment (joestringer) [inline pkg/bpf/map_linux.go]: I'd guess we only need to update this metric if `err != nil`, any particular reason not to have it inside the same defer function below which already has checks for whether the cache is enabled and checks against the `err`?

File: pkg/bpf/map_linux.go [hunk 13]
```diff
@@ -245,11 +257,51 @@ func (m *Map) scheduleErrorResolver() {
...
+		return
+	}
+
+	if !option.Config.MetricsConfig.BPFMapPressure {
+		if !m.withValueCache {
+			m.cache = nil
+		}
+		m.pressureGauge = nil
+		return
+	}
```
Comment (joestringer) [inline pkg/bpf/map_linux.go:303]: [0] Why would `m.pressureGauge` be non-nil at this point if `!option.Config.MetricsConfig.BPFMapPressure`? I would have thought that this field should be initialized to `nil` during the very first attempt to configure the gauge rather than checked & possibly disabled any time there is a map update.

I don't think we have any code that might cause `option.Config.MetricsConfig.BPFMapPressure()` to change at runtime so it should be safe to move it there.

File: pkg/endpoint/bpf.go [hunk 14]
```diff
@@ -1071,6 +1074,27 @@ func (e *Endpoint) SkipStateClean() {
...
+}
+
+func (e *Endpoint) updatePolicyMapPressureMetric() {
+	if e.policyMapPressureGauge == nil {
+		return
+	}
+
+	value := float64(len(e.realizedPolicy.PolicyMapState)) / float64(e.policyMap.MapInfo.MaxEntries)
+	e.policyMapPressureGauge.Set(value)
+}
```
Comment (joestringer) [inline pkg/endpoint/bpf.go:1101]: One of the reasons I was drawn to look at [0] was that I noticed this function appears to be equivalent but different in implementation due to the code I pointed out at [0]. Ideally these should be consistent (Ideally using the same object definition, shared across both users; but there's only a few lines so that's not such a big deal).

Would you mind either consistently doing the config check in the init() or in the update() across both implementations?

**Review by joestringer — CHANGES_REQUESTED** [review]
Thanks for working on this, and sorry about the delay in review :-)

Looks good overall, just a few minor nits here and there.

File: install/kubernetes/cilium/values.yaml (hunk 4 above)
Comment (jcaamano) [reply install/kubernetes/cilium/values.yaml]: Added a reference to the docs.

File: pkg/bpf/map_linux.go (hunk 9 above)
Comment (jcaamano) [reply pkg/bpf/map_linux.go:946]: `m.cache` is used both as a key-only cache for maps with pressure metric enabled but caching disabled, or as a key/value cache for maps with caching enabled. Thus `m.cache != nil` no longer tracks whether map caching is enabled, `withValueCache` does.

File: pkg/bpf/map_linux.go (hunk 10 above)
Comment (jcaamano) [reply pkg/bpf/map_linux.go:122]: Added comment on `m.cache` member that hopefully clarifies.

File: pkg/bpf/map_linux.go (hunk 11 above)
Comment (jcaamano) [reply pkg/bpf/map_linux.go:1006]: It didn't. Added.

File: pkg/bpf/map_linux.go (hunk 12 above)
Comment (jcaamano) [reply pkg/bpf/map_linux.go]: I removed the unnecessary `defer` and added `updatePressureMetric()` to the existing func. Still called when `err != nil` and caching enabled since even in that case an entry is added to the cache.
Also, I was calling `updatePressureMetric()` pretty liberally due to its idempotency but I tried to tweak better the places where it is called.

File: pkg/bpf/map_linux.go (hunk 13 above)
Comment (jcaamano) [reply pkg/bpf/map_linux.go:303]: On one side we cannot check `option.Config.MetricsConfig.BPFMapPressure` on initialization due to the static initialization of the maps when `option.Config` has not been loaded yet.

`m.pressureGauge` is initialized non-nil. While it could be initialized nil I chose to do it similar and inline with `m.cache` which cannot be initialized nil and which actually allows for less checks.

File: pkg/endpoint/bpf.go (hunk 14 above)
Comment (jcaamano) [reply pkg/endpoint/bpf.go:1101]: Here is different because we can check `option.Config.MetricsConfig.BPFMapPressure` on initialization which saves doing extra checks in the update.

File: pkg/bpf/map_linux.go (hunk 13 above)
Comment (joestringer) [reply pkg/bpf/map_linux.go:303]: Ah, static initialization and argument parsing strikes again. Thanks for the explanation.

This kind of detail would actually be nice to leave as a comment above the option check to ensure that in future we don't just assume we can move it and rely on the results (unless some other sync mechanism is put in place to ensure the options are available).

**Review by joestringer — APPROVED** [review]
Would be nice to add the explanation for why the check is there on every metric update into a comment but otherwise LGTM :+1:

File: pkg/bpf/map_linux.go (hunk 13 above)
Comment (jcaamano) [reply pkg/bpf/map_linux.go:303]: Added comment!

**Review by vadorovsky — APPROVED** [review]
ACK for @jcaamano's changes

**Comment by jrajahalme** [comment]
Docker rate limit hit in Travis CI, restarted.

**Comment by jrajahalme** [comment]
test-me-please

**Review by aanm — COMMENTED** [review]
@jcaamano the CI is failing and the errors are related with eBPF maps. I don't think these are flakes.

**Comment by jcaamano** [comment]
> @jcaamano the CI is failing and the errors are related with eBPF maps. I don't think these are flakes.

@aanm I took a look at this issue. The problem is triggered by removal of entries from `cilium_snat_v4_external` map that were never inserted in the first place. I don't know if this is expected or not, but I don't think it is caused by any change on this PR. The removal of these nonexistent entries causes a ENOENT error from the syscall and schedules the error resolver. The error resolver is triggered at a time when the map is closed and there is no valid fd to the map so when these entries are attempted to be removed again it causes EINVAL error to the syscall preventing the error resolver to complete successfully.

I have added the missing open to the error resolver. Now the error resolver gets ENOENT error from the syscall which was already accepted as a valid outcome and completes successfully. I don't know whether considering ENOENT an error in the original removal vs not in the error resolver is something that should be considered.

Heads up @joestringer.

**Comment by brb** [comment]
> The removal of these nonexistent entries causes a ENOENT error from the syscall and schedules the error resolver.

@jcaamano https://github.com/cilium/cilium/issues/11485 might be related.

**Comment by jcaamano** [comment]
> > The removal of these nonexistent entries causes a ENOENT error from the syscall and schedules the error resolver.
> 
> @jcaamano #11485 might be related.

Thanks for that reference @brb. Towards #11485 it might be worth it to consider ENOENT as a non-error outcome to the 'Delete' operation and perhaps adding an additional boolean to the return indicating if an entry was really deleted or not that the caller can check optionally. I do prefer it vs and additional call parameter. For the metric itself, I think the criteria should be fixed and the same for `Delete` and `resolveErrors`, whether it is to measure it or not.

Regarding the PR, I think that if adding the `open` call to `resolveErrors` looks sensible then we are good. But I cannot explain how this was not happening before unless it was what is being reported in #11485.

File: pkg/bpf/map_linux.go (hunk 13 above)
Comment (christarazi) [reply pkg/bpf/map_linux.go:303]: Is this PR (https://github.com/cilium/cilium/pull/14607) related? Is this what's meant by "static initialization"?

**Comment by mrostecki** [comment]
@jcaamano There are conflicts, please rebase :)

@brb Any thoughts on what @jcaamano's reply?

**Comment by brb** [comment]
If the test failures are due to #11485, then it's fine to ignore them in this PR.

**Comment by pchaigno** [comment]
test-me-please

**Comment by joestringer** [comment]
GKE cluster selection failed.

**Comment by joestringer** [comment]
retest-gke

**Comment by pchaigno** [comment]
All team reviews are covered, we have a good number of reviews, and CI is green. I'm marking as ready to merge.
